contracts/gasToken/GasTokenImplementationLogic.sol

```diff
@@ -0,0 +1,184 @@
+pragma solidity ^0.5.13;
+import "./ICHITOKEN.sol";
+import "../RelayerManager.sol";
+import "../libs/SafeMath.sol";
+import "../libs/Ownable.sol";
+
+contract GasTokenImplementationLogic is Ownable(msg.sender) {
+    using SafeMath for uint256;
+
+    ICHITOKEN public chiToken;
+    RelayerManager public relayerManager;
+
+    event Efficiency(uint256 indexed efficiency, uint256 indexed tokens);
+
+    // MODIFIERS
+    modifier onlyRelayerOrOwner() {
+        require(
+            relayerManager.getRelayerStatus(msg.sender) || msg.sender == owner(),
+            "You are not allowed to perform this operation"
+        );
+        _;
+    }
+
+    /**
+        @param destination contract address
+        @param data call data to be forwarded
+        @param mintPrice price at which gasToken was minted
+        @param gasLimit gas to execute the transaction
+        @dev Method forward the incoming transaction and then execute the gastoken functionality
+     */
+    function forward(
+        address payable destination,
+        bytes memory data,
+        uint256 mintPrice,
+        uint256 gasLimit
+    ) public onlyRelayerOrOwner {
+        uint256 initialGas = gasleft();
+
+        require(
+            executeCall(destination, data, gasLimit),
+            "ExecuteCall() failed"
+        );
+        refundGas(initialGas, mintPrice);
+    }
+
+    /**
+        @param to: destination contract address
+        @param data call data to be forwarded:
+        @dev Method execute the internal transaction send by user
+     */
+
+    function executeCall(
+        address to,
+        bytes memory data,
+        uint256 gasLimit
+    ) private returns (bool success) {
+        require(
+            gasleft() > ((gasLimit + (gasLimit / 63)) + 1000),
+            "Not enough gas"
+        );
+        assembly {
+            let txGas := gas
+            if iszero(eq(gasLimit, 0)) {
+                txGas := gasLimit
+            }
+            success := call(txGas, to, 0, add(data, 0x20), mload(data), 0, 0)
+        }
+    }
+
+    /**
+        @param initialGas: Gas sent by transaction at entry in forward method
+        @param mintPrice: price at which gasToken was minted
+        Calculate the total token needs to be burn and the efficiency.
+        if efficiency is >100 then only tokens will be burned to refund the gas
+     */
+    function refundGas(uint256 initialGas, uint256 mintPrice) private {
+        uint256 mintBase = 32254;
+        uint256 mintToken = 36543;
+        uint256 freeBase = 14154;
+        uint256 freeToken = 6870;
+        uint256 reimburse = 24000;
```

Comment:
This is amount of gas that can be refunded by burning 1 chi token... Here do the transaction that burn 1 chi token only... compare the value you get from doing gas estimation of same transaction... you'll get how much max gas was reimbursed by burning 1 chi token.